crengine/src/lvtinydom.cpp

```diff
@@ -7478,7 +7478,6 @@ ldomDocumentWriter::~ldomDocumentWriter()
         _document->getRootNode()->initNodeFont();
         //if ( !_document->validateDocument() )
         //    CRLog::error("*** document style validation failed!!!");
-        _document->updateRenderContext();
```

Comment:
Can you tell me a bit more of why/when that causes an issue ?
Could it be something that affects CoolReader only, and not KOReader, because we might do things a bit differently ?
(Just asking, I haven't really looked at why that's done here - but I haven't noticed any issue, or if I did I fixed any elsewhere.)

Reply:
Try this file from issue #140. 
```fb2
<?xml version="1.0" encoding="utf-8"?>
<FictionBook xmlns="http://www.gribuser.ru/xml/fictionbook/2.0" xmlns:l="http://www.w3.org/1999/xlink">
<stylesheet type="text/css">
  style[name="razryadka"] {
  letter-spacing: 0.4em;
  color: red;
  }
</stylesheet>

<p>One, <style name="razryadka">two</style>, three.</p>

</FictionBook>
```
If word ```two``` not red, you affected.

Reply:
It doesn’t even look right logically. We just loaded the file, but haven't rendered yet, but we keep the style hashes as after rendering.

Reply:
OK, so I'm affected too :) 
But your fix cause a full re-rendering to always be done after loading it seems.
With 3 of my usual EPUB documents, I get this - and I don't think they should or need that re-rendering:
```
08/01/20-13:54:35 DEBUG CreDocument: loading document...
08/01/20-13:54:39 DEBUG CreDocument: loading done.
08/01/20-13:54:39 DEBUG CreDocument: rendering document...
CRE: styles re-init needed after load, re-rendering
08/01/20-13:54:46 DEBUG CreDocument: rendering done.
```
So, doubling the load time.

It seems I get no issue and see it red with this in some HTML file:
```html
<html>
<head>
<style>
span[name=toto] { color: red; }
</style>
</head>
<body>
<H1>Margin vs <span name=toto>padding</span> vs borders</H1>
```

It fails also with just `<stylesheet type="text/css"> b { color: blue; } </stylesheet>` on the first loading - changing font size to have a re-rendering make it work.

So, the issue might more probably be with the `<stylesheet>` specific handling in FB2 (I think I remember it was specific - and it might be some spaghettugly stuff...) - that may be I've just never updated to work better. In the worst case we could just force that re-rendering when we meet such `<stylesheet>`. I'll have a look later.

Reply:
> We just loaded the file, but haven't rendered yet, but we keep the style hashes as after rendering.

It's an original crengine optimisation:

On first loading, when parsing the HTML, we compute styles of a node as we meet/enter it (because we can: we know its style and the style of its parent for inheritance) - and we compute rendering method when leaving the node (as then, we know all the rendering methods of its children). That way, after `~ldomDocumentWriter()`, styles and rendering methods are set - and we can just go at rendering (lvrend.cpp).

On re-rendering, the XHTML is valid, but we have to re-initNodeStyles (recursively, top first) and re-initNodeRendMethod (recursively, deep first) - and then, do the rendering (lvrend.cpp).

This double behaviour has always been painful, and one needs to check changes work in both codepaths (as you see with that FB2 `<stylesheet>`, it doesnt work in the 1st path). So, painful, but a nice optimisation I never dared dropping :)

Reply:
> If word two not red, you affected.

If you have some older CoolReader version (before any pick of my stuff), can you check if this ever worked on first load?
(Just so I know if I broke it - or if it was initially overlooked)

Reply:
I guess I have broken it with https://github.com/koreader/crengine/commit/9cecd154b where I commented out in `ldomDocumentWriter::OnTagClose()` what seems needed with FB2:` if isStyleSheetTag => document->applyDocumentStyleSheet();`
`This is now dealt with in :OnTagBody(), just before creating this <stylesheet> tag` was fine with HTML/EPUB - but with FB2, there is no body and it has to be done on the closing of the <stylesheet> tag as the real content is its sibling - and not children.

Reply:
This seems the right way to handle it:
```diff
--- a/crengine/src/lvtinydom.cpp
+++ b/crengine/src/lvtinydom.cpp
@@ -7629,6 +7629,7 @@ void ldomDocumentWriter::OnTagClose( const lChar16 *, const lChar16 * tagname )
         }
     }
     */
+    bool isStyleSheetTag = tagname[0] == 's' && tagname[1] == 't' && _currNode && !lStr_cmp(tagname, "stylesheet");

     lUInt16 id = _document->getElementNameIndex(tagname);
     //lUInt16 nsid = (nsname && nsname[0]) ? _document->getNsNameIndex(nsname) : 0;
@@ -7643,11 +7644,25 @@ void ldomDocumentWriter::OnTagClose( const lChar16 *, const lChar16 * tagname )
         _parser->Stop();
     }

-    /* This is now dealt with in :OnTagBody(), just before creating this <stylesheet> tag
-    if ( isStyleSheetTag ) {
+    // For EPUB/HTML, this is now dealt with in :OnTagBody(), just before creating this <stylesheet> tag.
+    // But for FB2, where we have:
+    //   <FictionBook>
+    //     <stylesheet type="text/css">
+    //       some css
+    //     </stylesheet>
+    //     <p>...
+    //     other content
+    //   </FictionBook>
+    // we need to apply the <stylesheet> content we have just left, so it applies
+    // to the coming up content.
+    // We check the parent we have just pop'ed is a <FictionBook>.
+    // Caveat: any style set on the <FictionBook> element itself won't be applied now
+    // in this loading phase (as we have already set its style) - but it will apply
+    // on re-renderings.
+    if ( isStyleSheetTag && _currNode->getElement()->getNodeId() == el_FictionBook ) {
         //CRLog::trace("</stylesheet> found");
 #if BUILD_LITE!=1
-        if ( !_popStyleOnFinish ) {
+        if ( !_popStyleOnFinish && _document->getDocFlag(DOC_FLAG_ENABLE_INTERNAL_STYLES) ) {
             //CRLog::trace("saving current stylesheet before applying of document stylesheet");
             _document->getStyleSheet()->push();
             _popStyleOnFinish = true;
@@ -7655,7 +7670,6 @@ void ldomDocumentWriter::OnTagClose( const lChar16 *, const lChar16 * tagname )
         }
 #endif
     }
-    */

     //logfile << " !c!\n";
 }
```
with the caveat mentionned in the comment - not sure if that's critical.
Also, there was/is some `#if BUILD_LITE!=1` in there - in the other stuff I did in the referenced commit above, I think I didn't wrap anything in `#if BUILD_LITE!=1` - so it's a bit inconsistent.
I dunno if CoolReader is still built somewhere with `BUILD_LITE=1` - but if not, you could remove it there - but if yes, it might need a bit more wrapping in a few other places.

Reply:
Just found some free time. Did you just return half of what you commented out? :) Ok it works, but you puts too much checks for not null value of _currNode, this check is at the begin of function onTagClose(). As I remember, you don't like unnecessary checks :)
Thank you for patch.

Reply:
> Did you just return half of what you commented out? :) 

Yep, I wasn't much aware of the FB2 stuff at the time :)

> you puts too much checks for not null value of _currNode, this check is at the begin of function onTagClose(). 

Indeed, feel free to clean that up.

I prefer my `bool isStyleSheetTag = tagname[0] == 's' && tagname[1] == 't' && _currNode && !lStr_cmp(tagname, "stylesheet");` just because we might meet a lot of other tags that start with "s" like `<span>`, and checking for "t" avoids the more expensive `Str_cmp()` on all of them.